vnc/vnc_clip.c

```diff
@@ -0,0 +1,1102 @@
+ * this to redirect virtual channel PDUs to this module */
+struct vnc_clipboard_data
+{
+    struct stream *rfb_clip_s;
+    int requested_clip_format; /* Last requested text format */
+    int active_data_requests; /* Number of outstanding FORMAT_DATA_REQUESTs */
+    struct stream *dechunker_s; /* Dechunker for the RDP clip channel */
+    int capability_version;  /* Clipboard virt channel extension version */
+    int capability_flags;    /* Channel capability flags */
+    bool_t startup_complete; /* is the startup sequence done (1.3.2.1) */
+};
+
+/**
+ * Summarise a stream contents in a way which allows two streams to
+ * be easily compared
+ */
+struct stream_characteristics
+{
+    char digest[16];
+    int length;
+};
+
+#ifdef USE_DEVEL_LOGGING
+/***************************************************************************//**
+ * Convert a CF_ constant to text
+ *
+ * @param CF_xxx constant
+ * @param buff Scratchpad for storing a temporary string in
+ * @param bufflen Length of the above
+ *
+ * @return string representation
+ */
+static const char *
+cf2text(int cf, char *buff, int bufflen)
+{
+    const char *result;
+
+    switch (cf)
+    {
+        case CF_UNICODETEXT:
+            result = "CF_UNICODETEXT";
+            break;
+
+        case CF_LOCALE:
+            result = "CF_LOCALE";
+            break;
+
+        case CF_TEXT:
+            result = "CF_TEXT";
+            break;
+
+        case CF_OEMTEXT:
+            result = "CF_OEMTEXT";
+            break;
+
+        default:
+            g_snprintf(buff, bufflen, "CF_<0x%08x>", cf);
+            result = buff;
+    };
+
+    return result;
+}
+#endif /* USE_DEVEL_LOGGING */
+
+/***************************************************************************//**
+ * Adds a CLIPRDR_HEADER ([MS-RDPECLIP] 2.2.1) to the data stream
+ *
+ * The location of the length is stored in the unused 'channel_hdr' field
+ * of the data stream. When send_stream_to_clip_channel() is called,
+ * we can use update the data length.
+ *
+ * @param s Output stream
+ * @param msg_type Message Type
+ * @param msg_flags Message flags
+ */
+static void
+out_cliprdr_header(struct stream *s, int msg_type, int msg_flags)
+{
+    out_uint16_le(s, msg_type);
+    out_uint16_le(s, msg_flags);
+    /* Save the datalen location so we can update it later */
+    s_push_layer(s, channel_hdr, 4);
+}
+
+/***************************************************************************//**
+ * Sends the contents of a stream buffer to the clipboard channel
+ *
+ * Stream contents are chunked appropriately if they are too big to
+ * fit in a single PDU
+ *
+ * The stream object cliprdr datalen header field is updated by this call.
+ *
+ * @param v VNC object
+ * @param s stream buffer
+ *
+ * @pre stream buffer must have been initialised with a call to
+ *      out_cliprdr_header()
+ * @pre Stream is terminated with s_mark_end()
+ */
+static int
+send_stream_to_clip_channel(struct vnc *v, struct stream *s)
```

Comment:
This chunking is specific to the RDP protocol and not the RFB protocol. Since you are changing the interface for the clipboard, why not move this RDP specific functionality to libxrdp, and have the converted but un-chunked stream as an argument to a new libxrdp "virtual channel send with chunking" function?

Reply:
It's a good point.

I actually started down this road, and then realised I was going to get in a bit of a mess with the dechunking coming in, as this needs to be done somewhere.

The thing that decided me on this route was that most of the channel data is handled by chansrv, not XRDP modules. If we do the chunking/dechunking in libxrdp, we'll end up sending giant messages between chansrv and xrdp which will result in noticeable pauses if (for example) the user decides to paste a few MB of data in one go. One of the advantages of chunking is it allows PDUs to be prioritised if necessary.

We could maybe provide a chunking/dechunking facility as part of libcommon, or even provide another library. What do you think?

Reply:
I agree that pipelining chunked messages between chansrv and the client is good for latency since the messages have to do an extra inter process communication hop.

I'm less convinced that there is a latency gain for chunking messages to the modules loaded in xrdp's memory space. It seems like the modules always need all chunks reassembled before any application logic can be run, and there is no cost for transferring a large stream to a loaded module.

If you think it's worth putting the chunking and de chunking logic into a library, I'd add it to libxrdp as it's own facility, and not integrated into the send/receive path.

Since you've already explored putting the chunking into libxrdp and determined that it's not the right path, I trust your judgement and we should go with the solution you've got.

Reply:
Thanks for that.

I agree that there's no great advantage for modules in re-assembling the chunks themselves. The SVC handler code in xrdp is pretty basic though, and if I'm reading it right (quite a big 'if') it just passes all SVC chunks for all channels through to a module if the module registers a handler in `mod_event`

https://github.com/neutrinolabs/xrdp/blob/96052ba6e3d97cfa6c0d41b039171d1db1a1eb59/xrdp/xrdp_wm.c#L1838-L1864

If we did the dechunking at that level we'd have to make the code more sophisticated to ignore chunks for unused channels, or we'd be re-assembling stuff just to bin it.

I'll raise an issue (tomorrow now) about this and identify all the chunking/dechunking code we're currently using.

Reply:
@aquesnel - I've raised #1920 as a separate issue regarding the chunking/dechunking.

Are you happy that satisfies the immediate requirement?

Feel free to add some detail to #1920 if you think it will be useful or necessary.

Reply:
That separate issue looks go to me, I think this PR is ready to ship.